### PR TITLE
Fix encoding of integers close to 4 bytes

### PR DESCRIPTION
**Pull request by prozacchiwawa**
I feel like this is a simpler way of expressing what was happening here that's more clear, and fixes the bug, new tests added.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -33,6 +33,8 @@ clvmr = "0.1.21"
 binascii = "0.1.4"
 yaml-rust = "0.4"
 linked-hash-map = "0.5.6"
+rand = "0.8.5"
+rand_chacha = "0.3.1"
```
Comment (arvidn) [inline Cargo.toml:39]: dependencies that are only required for tests can be added under `[dev-dependencies]` so they won't be linked against for production builds

File: src/classic/clvm/casts.rs [hunk 2]
```diff
@@ -125,6 +125,7 @@ pub fn bigint_to_bytes(v_: &Number, option: Option<TConvertOption>) -> Result<By
     let div = if signed { bi_one() } else { bi_zero() };
     let b32: u64 = 1_u64 << 32;
     let bval = b32.to_bigint().unwrap();
+    let (_sign, u32_digits) = v.to_u32_digits();
```
Comment (arvidn) [inline src/classic/clvm/casts.rs]: it seems to me that almost everything in this function could be replaced by a call to `v_.to_signed_bytes_be()` + a fix-up to prepend a leading 0 in case it's negative.

See this function: https://github.com/Chia-Network/clvm_rs/blob/main/src/number.rs#L8

File: src/compiler/sexp.rs [hunk 3]
```diff
@@ -23,6 +28,70 @@ pub enum SExp {
     Atom(Srcloc, Vec<u8>),
 }
 
+pub fn random_atom_name<R: Rng + ?Sized>(rng: &mut R, min_size: usize) -> Vec<u8> {
```
Comment (arvidn) [inline src/compiler/sexp.rs:36]: this looks like a function only used for tests. is it?
if so, it should have `#[cfg(test)]` above it.

also `random_sexp`, `random_atom` and `random_sexp`.

File: src/tests/classic/run.rs [hunk 4]
```diff
@@ -256,3 +264,90 @@ fn test_forms_of_destructuring_allowed_by_classic_1() {
...
+#[test]
+fn test_num_encoding_just_less_than_5_bytes() {
+    let res = do_basic_run(&vec!["run".to_string(), "4281419728".to_string()])
+        .trim()
+        .to_string();
+    assert_eq!(res, "0x00ff3147d0");
+}
+
+#[test]
+fn test_divmod() {
```
Comment (arvidn) [inline src/tests/classic/run.rs:292]: technically this is testing `div`, not `divmod`

File: src/tests/classic/run.rs [hunk 5]
```diff
@@ -256,3 +264,90 @@ fn test_forms_of_destructuring_allowed_by_classic_1() {
...
+    .trim()
+    .to_string();
+    assert_eq!(res, "18443");
+}
+
+struct RandomClvmNumber {
+    intended_value: Number,
+}
+
+fn random_clvm_number<R: Rng + ?Sized>(rng: &mut R) -> RandomClvmNumber {
```
Comment (arvidn) [inline src/tests/classic/run.rs]: this also looks like a test-only function.

**Review by arvidn — COMMENTED** [review]
there's a commit message saying it's adding fuzz targets, you didn't intend to do that in this PR, right?

File: Cargo.toml (hunk 1 above)
Comment (prozacchiwawa) [reply Cargo.toml:39]: I will change the imports.  I did intend to add the fuzzing.  It's been neglected a long time and would have prevented a lot of grief.  Making an effort to change my ways.

File: src/classic/clvm/casts.rs (hunk 2 above)
Comment (prozacchiwawa) [reply src/classic/clvm/casts.rs]: Thanks, I hadn't seen that part, but will simplify it.

File: src/compiler/sexp.rs (hunk 3 above)
Comment (prozacchiwawa) [reply src/compiler/sexp.rs:36]: Ya, i should get in that habit.

Comment (prozacchiwawa) [reply src/compiler/sexp.rs:36]: Although i have built bins that use these as well in the full fuzz branch, so they'll come right back eventually.

File: src/tests/classic/run.rs (hunk 4 above)
Comment (prozacchiwawa) [reply src/tests/classic/run.rs:292]: There's a macro named '/' which I added when there was controversy about this.  / was supposed to be deprecated.

File: src/tests/classic/run.rs (hunk 5 above)
Comment (prozacchiwawa) [reply src/tests/classic/run.rs]: ya will cfg it out.

File: src/classic/clvm/casts.rs (hunk 2 above)
Comment (prozacchiwawa) [reply src/classic/clvm/casts.rs]: It passes all the tests i have, but it was used unsigned in one place; as_path().  I _think_ that encoding is fine but need to write more tests.

File: src/compiler/sexp.rs (hunk 3 above)
Comment (arvidn) [reply src/compiler/sexp.rs:36]: you can condition it to the fuzzing configuration as well.
I usually get warnings for unused functions if I don't mark up functions with the right configuration. Maybe because these are `pub` they become part of the external interface and always considered used.

File: src/tests/classic/run.rs (hunk 4 above)
Comment (arvidn) [reply src/tests/classic/run.rs:292]: ah, right

File: src/classic/clvm/casts.rs (hunk 2 above)
Comment (prozacchiwawa) [reply src/classic/clvm/casts.rs]: added a lot more testing.

Comment (arvidn) [reply src/classic/clvm/casts.rs]: it seems to me that you're making it a lot harder for yourself by converting the bigint to `u32_digits()` rather than big endian bytes directly. it makes me a bit nervous because of the increased complexity and the increased number of edge cases that needs to be tested.

You say this is used for serializing a node path, which are always unsigned and never needs a leading 0 byte to make them unsigned, right?

If so, I can think of a few simplifications that would both make this function smaller but primarily bring down the number of test cases you need.

1. remove all handling of negative numbers in here, and just `assert!()` that the number is positive.
2. You could use a `BigUint` to represent the number, but I don't think it really makes a difference when converting to bytes. I think in this case even using `BigInt::to_be_bytes()` doesn't add a leading 0 for numbers that otherwise would be negative. So, as far as I understand, that's exactly what you want here.
3. Perhaps the function name should suggest that it's only for paths, or at least unsigned numbers
4. remove the `options` parameter

File: src/classic/clvm/casts.rs [hunk 6]
```diff
@@ -107,6 +107,8 @@ pub fn bigint_from_bytes(b: &Bytes, option: Option<TConvertOption>) -> Number {
     unsigned
 }
 
+// This is the version i ported from the typescript code re: chiaminejp.  It is
+// still used in some contexts, such as node_paths.
 pub fn bigint_to_bytes(v_: &Number, option: Option<TConvertOption>) -> Result<Bytes, String> {
     let v = v_.clone();
```
Comment (arvidn) [inline src/classic/clvm/casts.rs]: This seems pointless. The reference you take as a parameter, `v_` is immutable, and making a copy to an immutable variable doesn't seem to buy you anything. You might as well just use `v_` (and rename it ti `v`)

Comment (prozacchiwawa) [reply src/classic/clvm/casts.rs]: I've been in the habit of this in a few languages, mainly because i dislike the leaking of mutability (how it'll be used) into its declaration.  It's a style choice.
Ok will change it.

Comment (prozacchiwawa) [reply src/classic/clvm/casts.rs]: it used to be mutable, but along the way stopped needing that.

File: src/classic/clvm/casts.rs (hunk 2 above)
Comment (prozacchiwawa) [reply src/classic/clvm/casts.rs]: Apologies, i should have done much more up front porting this.  I reintroduced it after failing a recompile check without.
I am constantly trying to balance being responsive to people, getting things improved and delivering stuff and have cut that choice wrong a lot of time.

I put this back (it was here before) mainly to reduce the size of what i'm doing in this patch and hopefully to avoid creating a lot more work, but probably trying to save in the wrong place.

Let me look at rewriting it, apologies.

Comment (prozacchiwawa) [reply src/classic/clvm/casts.rs]: It's also the case that the rest of the code is organized around this, causing more churn.
I'll endeavor to separate out the uses.

Comment (prozacchiwawa) [reply src/classic/clvm/casts.rs]: rereading the node path code, i need to make sense of it a bit more and rewrite a chunk of it as well.  was hoping to avoid making too many changes in here (and potentially make things worse).

Comment (prozacchiwawa) [reply src/classic/clvm/casts.rs]: the option is actually important here to the way nodepath operates (at least in the ported form).  I see what the typescript code was meant to accomplish ... it normalizes the integer representation by first doing a signed conversion and then reading as unsigned.  the final conversions out when being written are unsigned.

Comment (prozacchiwawa) [reply src/classic/clvm/casts.rs]: Ok, presuming this version passes all checks, i think that's what you're looking for.

File: src/classic/clvm/casts.rs [hunk 7]
```diff
@@ -107,89 +106,39 @@ pub fn bigint_from_bytes(b: &Bytes, option: Option<TConvertOption>) -> Number {
...
-            (((1 << (8 * byte4_remain)) - lastbytes - dec) >> (8 * idx)) as u8
-        } else {
-            (lastbytes >> (8 * idx)) as u8
+// Pulled in code from clvm_rs to replace this function, re: PR comments.
+// The whole function relies on allocator so this is just the brass tacks.
+// Conversion type was here for completeness in the original code I ported
+// from the typescript.
+//
+// The unsigned option is easier, as used in as_path.
+pub fn bigint_to_bytes_clvm(v: &Number) -> Result<Bytes, String> {
```
Comment (arvidn) [inline src/classic/clvm/casts.rs]: this function cannot fail. It would simplify things if it would just return the `Bytes`.

I think this implementation is correct, but it would be nice to have it be demonstrated by unit tests.

`0` is an interesting case for instance, to ensure it does in fact produce an empty array.

File: src/classic/clvm/casts.rs [hunk 8]
```diff
@@ -107,89 +106,39 @@ pub fn bigint_from_bytes(b: &Bytes, option: Option<TConvertOption>) -> Number {
     unsigned
 }
 
-pub fn bigint_to_bytes(v_: &Number, option: Option<TConvertOption>) -> Result<Bytes, String> {
-    let v = v_.clone();
-
-    if v == bi_zero() {
+// Convert to bytes with no leading zeros added.  This allows numbers to appear
+// negative in the result stream but are treated as positive (for node paths).
+pub fn bigint_to_bytes_unsigned(v: &Number) -> Result<Bytes, String> {
```
Comment (arvidn) [inline src/classic/clvm/casts.rs]: I don't see unit tests for this function either. having tests demonstrating the difference between this and `bigint_to_bytes_clvm` would be nice.

on an unrelated note, I'm a bit surprised to see `Bytes` be an optional type, and that an empty array is represented as `None`. Is this an artifact from a direct translation from Python or JS?

**Comment by prozacchiwawa** [comment]
I've been putting tests for these into run.rs, testing them in a wholistic
way since this is buried, but makes sense to add tests for just these as
well i suppose.  I wanted to ensure that compilation as a whole produced
the right result, and there are layers above.

Bytes is directly translated from the typescript interpretation of python's
bytes by chiaminejp.  It's a bit more ugly but models the use of undefined
in ts.

File: src/tests/classic/run.rs [hunk 9]
```diff
@@ -256,3 +279,206 @@ fn test_forms_of_destructuring_allowed_by_classic_1() {
...
+
+        // We'll have it compile a constant value.
+        // The representation of the number will come out most likely
+        // as a hex constant.
+        let serialized_through_run = do_basic_run(&vec![
+            "run".to_string(),
+            format!("(q . {})", number_spec.intended_value),
+        ])
+        .trim()
+        .to_string();
```
Comment (arvidn) [inline src/tests/classic/run.rs:356]: this test isn't quite as good as a unit test. it's not targeting edge cases (e.g. byte-boundary values). It also "launders" possible encodings with redundant leading zeroes or `\xff` bytes, since you just compare the value after it's interpreted as an integer again.

If it fails, it's also hard to reproduce since the random sequence is different for every run.

File: src/tests/compiler/clvm.rs [hunk 10]
```diff
@@ -138,3 +150,66 @@ fn test_clvm_4() {
...
+        let number_spec: RandomClvmNumber = rng.gen();
+
+        let to_bytes_clvm = bigint_to_bytes_clvm(&number_spec.intended_value).raw();
+        let to_bytes_unsigned = if number_spec.intended_value < bi_zero() {
+            None
+        } else {
+            Some(bigint_to_bytes_unsigned(&number_spec.intended_value).raw())
+        };
+
+        if number_spec.intended_value == bi_zero() {
```
Comment (arvidn) [inline src/tests/compiler/clvm.rs:179]: this case may not run. it's hard for me to judge how likely it is to run, but I would think we would always want to test this property.

File: src/tests/compiler/clvm.rs [hunk 11]
```diff
@@ -138,3 +150,66 @@ fn test_clvm_4() {
...
+            if let Some(usbi) = &to_bytes_unsigned {
+                assert!(usbi.is_empty());
+            }
+            continue;
+        }
+
+        // Determine whether an extension byte would be needed.
+        let need_ext_byte = does_number_need_extension_byte(number_spec.intended_value.clone());
+        if need_ext_byte {
+            assert_eq!(to_bytes_clvm[0], 0);
```
Comment (arvidn) [inline src/tests/compiler/clvm.rs:190]: Given that the test cases are random, are you sure this case is executed?

**Review by arvidn — APPROVED** [review]
I'm not super excited about testing random values. I think they are unlikely to cover the interesting cases.

File: src/tests/compiler/clvm.rs (hunk 10 above)
Comment (prozacchiwawa) [reply src/tests/compiler/clvm.rs:179]: quite a few explicit values are tested elsewhere in the stack via their run outcomes.
i've added randomized ones because i didn't capture all the right ones in those and figured that this would do better.
the specific one this is in run.rs at line 292.

**Comment by prozacchiwawa** [comment]
I'll take this for now and round on it again i suppose when the time is there.  maybe i should have kept at my ill fated foray into tdd that left me far behind and had everyone pissed off.
